### ISSUE TITLE
revdep hesim object not found when re-building vignettes

### ISSUE THREAD

**Issue by tdhock** [Rdatatable/data.table#5541]
revdep hesim has a new error when re-building vignettes with data table from github master installed,
```
* checking re-building of vignette outputs ... ERROR
Error(s) in re-building vignettes:
--- re-building 'cea.Rmd' using rmarkdown
Quitting from lines 271-283 (cea.Rmd) 
Error: processing vignette 'cea.Rmd' failed with diagnostics:
object 'w' not found
```
git bisect says this error started happening after merging this commit https://github.com/Rdatatable/data.table/commit/9619ad5fcde4fc951f8aed8b34bdc5bd61dd6ef5

**Comment by jangorecki** [comment]
Logic in linked commit looks like a good place for a helper function.

**Comment by tdhock** [comment]
This is the code with the issue https://github.com/hesim-dev/hesim/blob/main/vignettes/cea.Rmd

**Comment by tdhock** [comment]
line of code with error is
```r
> totevpi <- evpi[,lapply(.SD, weighted.mean, w = w),
+ by = "k", .SDcols = c("evpi")]
Error in eval(jsub[[ii]][[3L]], parent.frame()) : object 'w' not found
> evpi
Key: <grp>
         grp      k       best        enmbci        enmbpi          evpi     w
      <char>  <num>     <char>         <num>         <num>         <num> <num>
  1: Group 1      0 Strategy 1 -7.427476e+00 -7.427476e+00  9.769963e-15  0.25
  2: Group 1    500 Strategy 1  3.994839e+03  3.994839e+03 -2.273737e-12  0.25
  3: Group 1   1000 Strategy 1  7.997105e+03  7.997105e+03  5.456968e-12  0.25
  4: Group 1   1500 Strategy 1  1.199937e+04  1.199937e+04 -1.455192e-11  0.25
  5: Group 1   2000 Strategy 1  1.600164e+04  1.600164e+04  0.000000e+00  0.25
 ---                                                                          
798: Group 2 198000 Strategy 3  2.117403e+06  2.156015e+06  3.861221e+04  0.75
799: Group 2 198500 Strategy 3  2.122903e+06  2.161614e+06  3.871051e+04  0.75
800: Group 2 199000 Strategy 3  2.128404e+06  2.167212e+06  3.880881e+04  0.75
801: Group 2 199500 Strategy 3  2.133904e+06  2.172811e+06  3.890711e+04  0.75
802: Group 2 200000 Strategy 3  2.139404e+06  2.178410e+06  3.900541e+04  0.75
```

**Comment by tdhock** [comment]
reverting that commit fixes that error, 
```r
> totevpi <- evpi[,lapply(.SD, weighted.mean, w = w),
+ by = "k", .SDcols = c("evpi")]
> totevpi
          k          evpi
      <num>         <num>
  1:      0 -4.884981e-15
  2:    500  2.501110e-12
  3:   1000  8.185452e-12
  4:   1500  4.547474e-12
  5:   2000  4.092726e-12
 ---                     
397: 198000  3.033478e+04
398: 198500  3.041163e+04
399: 199000  3.048848e+04
400: 199500  3.056533e+04
401: 200000  3.064218e+04
```

**Comment by tdhock** [comment]
TODOs
- add a test case similar to this that fails for current master.
- write a helper function to simplify the code in data.table.R?
- add a test case which captures the problem that https://github.com/Rdatatable/data.table/commit/9619ad5fcde4fc951f8aed8b34bdc5bd61dd6ef5 was supposed to fix?

**Comment by ben-schwen** [comment]
Seems similar to #5547, although here the problem is that `n` is not evaluated while there the problem is that `n` is evaluated in the wrong environment.

**Comment by ColeMiller1** [comment]
MRE for the revdep issue:

``` r
dt = data.table(id = rep(1:3, each = 10L), val = rnorm(30L), w = runif(30L))
dt[, lapply(.SD, weighted.mean, w), by = id, .SDcols = "val"]

## Error in eval(jsub[[ii]][[3L]], parent.frame()) : object 'w' not found
```
I can look at making a full test example this weekend with expected results.

**Comment by jangorecki** [comment]
If reverting NSE introduced in `shift` fixes the issue then I think it is the proper way to go.

**Comment by ben-schwen** [comment]
Actually 9619ad5 removed the check for existence which is now erroring. @jangorecki is there a way of triggering the Gitlab CI (testing for older version) without merging?

**Comment by tdhock** [comment]
I'm not sure I fully understand this issue, but because it involves weighted mean, it seems similar to #5543 which Ben asked revdep to fix, so is this something we should fix, or ask hesim to fix?

**Comment by jangorecki** [comment]
> Actually [9619ad5](https://github.com/Rdatatable/data.table/commit/9619ad5fcde4fc951f8aed8b34bdc5bd61dd6ef5) removed the check for existence which is now erroring. @jangorecki is there a way of triggering the Gitlab CI (testing for older version) without merging?

Yes, pushing a branch directly to gitlab repo.

**Comment by jangorecki** [comment]
Maybe this one will be fixed by https://github.com/Rdatatable/data.table/pull/5548 ?

**Comment by ben-schwen** [comment]
> Maybe this one will be fixed by #5548 ?

Yes, by altering the signature to `gweighted.mean = function(x, w, ..., na.rm=FALSE)` which is also closer to `stats:::weighted.mean.default = function (x, w, ..., na.rm = FALSE)`.

Also added a test that wrongly named arguments really get lost in `...`

**Comment by jangorecki** [comment]
Not just closer but exact match, something that seems good design decision
